citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java

```diff
@@ -979,6 +991,137 @@ protected void doPostJsonResponse(ParameterParser params,
 		return results;
 	}
 
+	protected Map<String, Object> addSection(ParameterParser params, SessionState state) {
+		String message;
+		Map<String, Object> results = new HashMap<String, Object>();
+		try {
+			int locationId = params.getInt("locationId");
+			CitationCollectionOrder.SectionType sectionType = CitationCollectionOrder.SectionType.valueOf(params.getString("sectionType"));
+			CitationCollection collection = getCitationCollection(state, false);
+
+			CitationCollectionOrder citationCollectionOrder = new CitationCollectionOrder(collection.getId(), locationId, sectionType, H1_SECTION_TITLE);
+			getCitationService().saveSection(citationCollectionOrder);
+			message = rb.getString("resource.updated");
+		}
+		catch (Exception e){
+			message = e.getMessage();
+			logger.warn("Exception in addSection() " + e);
+		}
+		if(message != null && ! message.trim().equals("")) {
+			results.put("message", message);
+		}
+		return results;
+	}
+
+	protected Map<String, Object> addSubSection(ParameterParser params, SessionState state) {
+		String message;
+		Map<String, Object> results = new HashMap<String, Object>();
+
+		try {
+			int locationId = params.getInt("locationId");
+			String addSectionHTML = params.getString("addSectionHTML");
+			CitationCollectionOrder.SectionType sectionType = CitationCollectionOrder.SectionType.valueOf(params.getString("sectionType"));
+			CitationCollection collection = getCitationCollection(state, false);
+
+			CitationCollectionOrder citationCollectionOrder = new CitationCollectionOrder(collection.getId(), locationId, sectionType, addSectionHTML);
+			getCitationService().saveSubsection(citationCollectionOrder);
+			message = rb.getString("resource.updated");
+		}
+		catch (Exception e){
+			message = e.getMessage();
+			logger.warn("Exception in addSubSection() " + e);
+		}
+
+		if(message != null && ! message.trim().equals("")) {
+			results.put("message", message);
+		}
+		return results;
+	}
+
+	protected Map<String, Object> dragAndDrop(ParameterParser params, SessionState state) {
+
+		String message = null;
+		Map<String, Object> results = new HashMap<String, Object>();
+		ObjectMapper mapper = new ObjectMapper();
+		List<CitationCollectionOrder> citationCollectionOrders;
+
+		try {
+			String citationCollectionId = params.getString("citationCollectionId");
+			String nestedCitations = params.getString("data");
+			if (nestedCitations!=null){
+
+				// remove extra parentheses in json
+				// needed because of the extra ol's for accordion effect
+				nestedCitations = nestedCitations
```

Comment:
Isn't this really brittle? Does this mean that layout changes in the HTML will need this section updating?